packages/js/product-editor/src/blocks/input/edit.tsx

```diff
@@ -0,0 +1,115 @@
+/**
+ * External dependencies
+ */
+import { createElement, createInterpolateElement } from '@wordpress/element';
+import type { BlockAttributes } from '@wordpress/blocks';
+import { useBlockProps } from '@wordpress/block-editor';
+import { useEntityProp } from '@wordpress/core-data';
+import { useInstanceId } from '@wordpress/compose';
+import { __ } from '@wordpress/i18n';
+import { Product } from '@woocommerce/data';
+import classNames from 'classnames';
+import {
+	BaseControl,
+	// @ts-expect-error `__experimentalInputControl` does exist.
+	__experimentalInputControl as InputControl,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { useValidation } from '../../contexts/validation-context';
+
+interface Metadata {
+	id?: number;
+	key: string;
+	value: string;
+}
+
+export function Edit( { attributes }: { attributes: BlockAttributes } ) {
+	const blockProps = useBlockProps();
+	const {
+		property,
+		label,
+		placeholder,
+		required,
+		validationRegex,
+		validationErrorMessage,
+	} = attributes;
+	const [ metadata, setMetadata ] = useEntityProp< Metadata[] >(
+		'postType',
+		'product',
+		'meta_data'
+	);
+
+	const value =
+		metadata.find( ( item ) => item.key === property )?.value || '';
+
+	const nameControlId = useInstanceId( BaseControl, property ) as string;
+
+	const { error, validate } = useValidation< Product >(
+		property,
+		async function validator() {
+			if ( required && ! value ) {
+				return __( 'This field is required.', 'woocommerce' );
+			}
+			if ( validationRegex ) {
+				const regExp = new RegExp( validationRegex );
+				if ( ! regExp.test( value ) ) {
+					return (
+						validationErrorMessage ||
+						__( 'Invalid value for the field.', 'woocommerce' )
+					);
+				}
+			}
+		},
+		[ value ]
+	);
+
+	return (
+		<div { ...blockProps }>
+			<BaseControl
+				id={ nameControlId }
+				label={
+					required
+						? createInterpolateElement( `${ label } <required />`, {
+								required: (
+									<span className="woocommerce-product-form__required-input">
+										{ __( '*', 'woocommerce' ) }
```

Comment:
We need a comment for the translators so they know how to handle this symbol.

Reply:
to be honest I'm not even sure this requires translation (maybe). I copied it from the 'name' field. But I'll add the comment for now.